diff -urN portaudio/configure.in portaudio.new//configure.in
--- portaudio/configure.in	2011-03-25 17:59:27.000000000 +0100
+++ portaudio.new//configure.in	2011-04-10 22:35:28.547814050 +0200
@@ -172,8 +172,6 @@
       [[ "$enable_mac_universal" = "yes" ] &&
        [ "$enable_mac_debug" != "yes" ]] ; then
    CFLAGS="-O2 -Wall -pedantic -pipe -fPIC -DNDEBUG"
-else
-   CFLAGS=${CFLAGS:-"-g -O2 -Wall -pedantic -pipe -fPIC"}
 fi
 
 if [[ $ac_cv_c_bigendian = "yes" ]] ; then
