--- Makefile.in~	2004-01-29 21:36:27.000000000 +0100
+++ Makefile.in	2005-06-06 22:43:49.834317600 +0200
@@ -140,9 +140,9 @@
 	$(CC) -o $@ $(CFLAGS) pa_tests/$*.c lib/$(PALIB) $(LIBS)
 
 install: lib/$(PALIB) lib/$(PADLLV)
-	$(INSTALL) -m 644 lib/$(PADLLV) $(PREFIX)/lib/$(PADLLV)
-	$(INSTALL) -m 644 lib/$(PALIB) $(PREFIX)/lib/$(PALIB)
-	cd $(PREFIX)/lib && rm -f $(PADLL) && ln -s $(PADLLV) $(PADLL)
+	$(INSTALL) -m 644 lib/$(PADLLV) $(PREFIX)/$(LIB)/$(PADLLV)
+	$(INSTALL) -m 644 lib/$(PALIB) $(PREFIX)/$(LIB)/$(PALIB)
+	cd $(PREFIX)/$(LIB) && rm -f $(PADLL) && ln -s $(PADLLV) $(PADLL)
 	$(INSTALL) -m 644 pa_common/portaudio.h $(PREFIX)/include/portaudio.h
 	@echo ""
 	@echo "------------------------------------------------------------"
@@ -151,7 +151,7 @@
 	@echo "On some systems (e.g. Linux) you should run 'ldconfig' now"
 	@echo "to make the shared object available.  You may also need to"
 	@echo "modify your LD_LIBRARY_PATH environment variable to include"
-	@echo "the directory $(PREFIX)/lib"
+	@echo "the directory $(PREFIX)/$(LIB)"
 	@echo "------------------------------------------------------------"
 	@echo ""
 
